### ISSUE TITLE
Map data is erroneously placed or removed independently of the cursor

### ISSUE THREAD

**Issue by Jackal-42** [Jackal-42/Industry-Game#3]
https://user-images.githubusercontent.com/91426054/141340061-7154223b-80f7-4c75-808b-c380e48688cf.mp4

I thought I knew the pipe function inside and out, and then this pops up. While I don't know for sure, I suspect the problem is an error in calculating where the pipe should be placed in relation to the total world map rather than in relation to the window.

**Comment by Jackal-42** [comment]
Side note: only happens when beginning or ending the mouse hold event. Important for debugging purposes

**Comment by robtech21** [comment]
Wait, this is exactly the issue I was having yesterday!

**Comment by Jackal-42** [comment]
I didn't realize it was the same one. Maybe I could merge or something. Hopefully I will find a quick fix?

**Comment by robtech21** [comment]
Current findings on the issue:

* Event only triggers upon `mouseup`
* -1 cursor Y change results in -2 ghost point Y change

Examples:
* (13, 28) and (17, 21)
* (9, 28) and (10, *21)
* (18, 27) and (27, 19)
* (16, 29) and (23, 23)

**Comment by Jackal-42** [comment]
I've investigated the points we have graphed already and found something interesting. The pipe deletion is NOT locked to the cursor's position, e.g. the deletion only happens when the mouse is in the top left corner of the box and not when it is anywhere else. Because the event seems to not be locked to the grid, I think that this pretty much confirms the theory that it is an error with the system for determining where the cursor is. And yes, it does only occur on mouse up. With this knowledge, I fully expect to have it fixed by tonight. But since when was it ever that straightforward? I suppose we may get lucky.

**Comment by Jackal-42** [comment]
And, I think I've fixed it. And it was literally two lines of code. (Of course, more debugging will be required in order to make sure it is fully stable. There is still at least one bug I am aware of which results in a crash) I will briefly explain what happened.

The "old" system for determining where the cursor was just used the position of the mouse in relation to the window. I implemented a new system when a scrolling map was added which made the cursor's position need to be adjusted in relation to the scroll as well. And there was one part of the algorithm I forgot to change, the part where, upon mouseup, it gets the position of the mouse and places a pipe there. (The reason for this is to connect the pipes near the place where the mouse is let up, I think it was leftover from a time when placing a pipe caused the pipes around them to update.) It was a bit of a code relic I thought was still essential. So, it placed the pipes in relation to the window and didn't account for the scroll. Also, there was a weird multiplier effect because in the new version I made the window size scale to fit the browser tab, and the code wasn't accounting for that. Anyway, I think it is fixed. Keep on finding bugs!
